packages/studio-base/src/panels/Plot/processor/downsample.ts

```diff
@@ -0,0 +1,757 @@
+  //
+  // this is just a guess, but mostly works, and if the plot gets too dense we
+  // will downsample again anyway
+  const newRange = getBoundsRange(newBounds);
+  if (previous == undefined || chunkSize === 0) {
+    const proportion = newRange / viewportRange;
+    if (proportion < minSize) {
+      // Before we can infer a reasonable bucket size, we still want to show
+      // data, so we fall back to the previous algorithm
+      const indices = downsampleTimeseries(iterateTyped(raw.data), view);
+      const resolved = resolveTypedIndices(raw.data, indices);
+      if (resolved == undefined) {
+        return state;
+      }
+
+      return {
+        ...state,
+        dataset: {
+          ...raw,
+          pointRadius: 0,
+          data: resolved,
+        },
+      };
+    }
+
+    const bestGuessBuckets = Math.min(
+      Math.floor((newRange / viewportRange) * maxPoints),
+      maxPoints,
+    );
+    const downsampled = downsampleDataset(newData, bestGuessBuckets);
+    if (downsampled == undefined) {
+      return state;
+    }
+
+    return {
+      ...state,
+      cursor: newCursor,
+      chunkSize: newCursor,
+      numBuckets: bestGuessBuckets,
+      dataset: {
+        ...raw,
+        pointRadius: 0,
+        data: downsampled,
+      },
+    };
+  }
+
+  // in order for the downsampled signal to maintain the same visual density
+  // over the entire plot, the number of points we downsample needs to stay the
+  // same; this is what `chunkSize` does.
+  //
+  // most of the time, however, we receive new points in quantities far smaller
+  // than `chunkSize`, so to get around this (but still retain visual density)
+  // we reuse raw points that already exist in the dataset and start our
+  // downsample a few buckets _before_ the points we're adding. this is because
+  // the point we choose for each bucket depends on the point we chose in the
+  // previous bucket, and so on.
+  //
+  // we then append any new _downsampled_ points to our accumulated downsampled
+  // dataset.
+  const numNewPoints = newCursor - oldCursor;
+  if (numNewPoints >= chunkSize) {
+    // the `chunkSize` is also the upper bound for the number of points we
+    // process in one go (again to maintain similar visual density)
+    const downsampled = downsampleDataset(sliceTyped(newData, 0, chunkSize), numBuckets);
+    if (downsampled == undefined) {
+      return state;
+    }
+
+    // we go around again and consume all the data we can
+    return updateSource(path, params, {
+      ...state,
+      cursor: oldCursor + chunkSize,
+      dataset: concatDataset(previous, { ...previous, data: downsampled }),
+    });
+  }
+
+  const numOldPoints = chunkSize - numNewPoints;
+  const rawStart = newCursor - chunkSize;
+  const pointsPerBucket = Math.trunc(chunkSize / numBuckets);
+  const lastRawBucket = Math.max(
+    (numOldPoints - (numOldPoints % pointsPerBucket)) / pointsPerBucket - 2,
```

Comment:
would this give the same result?
```suggestion
    Math.floor(numOldPoints / pointsPerBucket) - 2,
```